Feature branch for GEOPM service

cmcantalupo opened the pull request:
Creates a new directory with separate build infrastructure that supports the geopm systemd service.  This service will provide the PlatformIO features through a daemon interface with access control capabilities and also save/restore capabilities.

cmcantalupo committed:
Initial sketch of geopm service

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Quick conversion of xml to dasbus daemon

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add methods to query information about signals and controls

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

dannosliwcd commented on service/bin/geopmd:
```diff
@@ -0,0 +1,103 @@
+import geopmpy.pio
+
+class GEOPMService(object):
+    __dbus_xml__ = """
```
Looks like dasbus can also generate the xml for us if we use python3 type annotations and decorate the class with `@dasbus.dbus_interface`. Not sure if there are advantages/disadvantages to that approach though.

cmcantalupo replied:
We will probably need to have access to the XML to write a C++ client, that was my thinking behind creating the interface/io.github.geopm.xml file.

cmcantalupo replied:
TBD in feature branch...This seems like a good feature of dasbus to use.

dannosliwcd commented on service/bin/geopmd:
```diff
@@ -0,0 +1,103 @@
+    def AccessList(self, group):
+        return self._pio.signal_names(), self._pio.control_names()
+
+    def AccessControl(self, group,
```
Maybe `SetAccessList`

dannosliwcd commented on service/bin/geopmd:
```diff
@@ -0,0 +1,103 @@
+    def __init__(self, pio=geopmpy.pio):
+        self._pio = pio
+
+    def AccessList(self, group):
```
Maybe `GetAccessList`?

dannosliwcd commented on service/bin/geopmd:
```diff
@@ -0,0 +1,103 @@
+    def AccessDelete(self, group):
+        raise NotImplementedError
+
+    def AccessAllowed(self):
```
Do we need both `AccessAllowed` and `AccessList`? Looks like they are the same except this also gives domain context.

cmcantalupo replied:
My idea here was that AccessAllowed() is for clients to find out what they might want to add to the Loop call (hence no group input parameter, and the domain), where as the AccessList is for an administrator to get the access list for a specified group.  If we can (unclear to me) restrict the AccessList to be root only, it will make more sense.

dannosliwcd commented on service/bin/geopmd:
```diff
@@ -0,0 +1,103 @@
+                    for nn in self._pio.control_names()]
+        return signals, controls
+
+    def Loop(self, signals, controls, interval):
```
Maybe `StartXYZLoop`? I guess `XYZ` could be something like `PlatformIO`, `Sampling`, `Control`, `GeopmControl`, etc. Or maybe `StartGeopmController`

cmcantalupo replied:
Decided on a concept of opening and closing a "session".   LMKWYT

dannosliwcd commented on service/dbus-1/interfaces/io.github.geopm.xml:
```diff
@@ -0,0 +1,136 @@
+      <doc:doc>
+        <doc:description>
+          <doc:para>
+            Enables a client to query information about an array of signals.
```
Array of `signals`->`controls`

dannosliwcd commented on service/system/systemd-geopm.service:
```diff
@@ -0,0 +1,11 @@
+Description=User space interface to hardware features with fine grained access control
+
+[Service]
+Environment=PYTHONUNBUFFERED=true
```
This reminds me, do we need to do anything special for output, like using a systemd logger or something? Or is print-based logging sufficient?

cmcantalupo replied:
Good question, I've certainly seen special logging it in some examples for a variety of reasons.  We can look into it.   Messages are turning up in the syslog, but we may want our own log file.

dannosliwcd commented:
Should one of these methods (or an additional method) export a mapping of (signal/control)->(offset within shmem buffer)?

cmcantalupo committed:
Move required source files into service/src

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add a few missing BSD headers

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
First draft of configure.ac for service

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Get build working

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
[WIP] Get packaging together

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Adding the geompdpy module

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Clean up spec file

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Clean up .gitignore

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fixup packaging

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add missing files to spec

cmcantalupo committed:
Fix import in geompd

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Move to strictly python36 packaging

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Get rid of references to geopmpy

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Closer to getting the service running

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Name service properly

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Got the service up and running, having some issues with multiple return values from dbus methods

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Remove initial draft of D-Bus interface

- May bring back these APIs later, but we will build them up one by one.
- Leaving the Test interface in place for now, since it is the only one that works.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Refactor the service into an interface using D-Bus and an implementation that can be tested separately

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
[WIP] New sketch for PlatformService interface

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Bit of code clean up

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Clean up build and add geopmaccess

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
[WIP] Trying to wire up service

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Comment out code to get pid and uid from dbus until it is working

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Update interface xml document

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix up logic for getting user id and pid

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Service is working better

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Doing some debugging

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Try to restrict the set interface

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Rename calling_pid to client_pid

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Put together C++ and C APIs for opening and closing a session

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Use key to close session, not PID (will check PID in daemon)

- Bit of other clean up in place.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Start to add some documentation to the README

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

dannosliwcd commented on service/README.md:
```diff
@@ -0,0 +1,95 @@
+Each geopm signal and control has an associated name and a hardware
+domain.  The name is a unique string identifier defined by the IOGroup
+that provides the signal or control.  The hardware domain that
+provides the interface is represented by one of the enumerations as
```
Specifically, this is the integer value of those enumerations, right? (vs. a stringified enum identifier)

dannosliwcd replied:
Also, domain _index_ is never used at this scope of the interface, right? (No need to explicitly state that here, just checking in case I'm assuming wrong)

cmcantalupo replied:
correct

dannosliwcd commented on service/README.md:
```diff
@@ -0,0 +1,95 @@
+lists that are associated with groups that they belong to.  The default
+lists are stored in:
+
+    /etc/geopm-service/0.DEFAULT_ACCESS/allowed_signals
```
Does the zero have any meaning specific to us, or is that just a D-Bus or systemd convention?

cmcantalupo replied:
I just needed a number, because groups cannot begin with numbers.

cmcantalupo commented:
This PR was merged into the geopm-service branch of the main repository to facilitate collaboration.  There are still many issues both mentioned in the PR and also unmentioned, that must be addressed before the geopm-service branch will be merged to the dev branch.
